File: Java/linkedList/singlyLinked/SinglyLinkedLinkedList.java
package linkedList.singlyLinked;

public class SinglyLinkedLinkedList 
{
	private Node head;
	public void addToHead(int dataVal) 
	{
		Node temp=new Node();
		temp.setDataVal(dataVal);
		temp.setNext(null);
		if(head==null)
		{
			head=temp;
		}
		else
		{
			temp.setNext(head);
			head=temp;
		}
	}
	public void addNode(int dataVal)
	{
		Node temp=new Node();
		temp.setDataVal(dataVal);
		temp.setNext(null);
		if(head==null)
		{
			head=temp;
		}
		else
		{
			Node temphead=head;
			while(!(temphead.getNext()==null))
			{
				temphead=temphead.getNext();
			}
			temphead.setNext(temp);
		}
	}
	public void addAtIndex(int dataVal,int index)
	{
		if(index==0)
		{
			addToHead(dataVal);
		}
		else
		{
			Node temp=new Node(dataVal, null);
			int currIndex=0;
			Node temphead=head;
			while(temphead.getNext()!=null && index!=1)
			{
				temphead=temphead.getNext();
				currIndex++;
				if(currIndex==index-1)
				{
					break;
				}
			}
			temp.setNext(temphead.getNext());
			temphead.setNext(temp);
		}
	}
	public void updateAtIndex(int dataVal,int index)
	{
		if(index==0)
		{
			head.setDataVal(dataVal);
		}
		else 
		{
			Node temphead=head;
			int currIndex=0;
			while(temphead.getNext()!=null)
			{
				temphead=temphead.getNext();
				currIndex++;
				if(currIndex==index)
				{
					break;
				}
			}
			if(index>currIndex)
			{
				System.out.println("Out of Bounds");
			}
			else
			{
				temphead.setDataVal(dataVal);
			}
		}
	}
	public void deleteHead()
	{
		head=head.getNext();
	}
	public void deleteLast()
	{
		Node temphead=head;
		while(temphead.getNext().getNext()!=null)
		{
			temphead=temphead.getNext();
		}
		temphead.setNext(null);
	}
	public void deleteAtIndex(int index)
	{
		if(index==0)
		{
			deleteHead();
		}
		else
		{
			Node temphead=head;
			int currIndex=0;
			while(temphead.getNext()!=null && index!=1)
			{
				temphead=temphead.getNext();
				currIndex++;
				if(currIndex==index-1)
				{
					break;
				}
			}
			if(index>currIndex)
			{
				System.out.println("Out of Bounds");
			}
			else
			{
				temphead.setNext(temphead.getNext().getNext());
			}
		}
	}
	public int size()
	{
		int size=0;
		if(head!=null)
		{
			Node temphead=head;
			while(temphead.getNext()!=null)
			{
				temphead=temphead.getNext();
				size++;
			}
			size++;
		}
		return size;
	}
	public void printAllNodeValues()
	{
		String allVals="";
		Node temphead=head;
		while(temphead.getNext()!=null)
		{
			allVals+=temphead.getDataVal()+" , ";
			temphead=temphead.getNext();
		}
		allVals+=temphead.getDataVal();
		System.out.println(allVals);
	}
	public void reverseRecursive()
	{
		head=reverseRecursive(head);
	}
	private Node reverseRecursive(Node head)
	{
		if(head==null||head.getNext()==null)
		{
			return head;
		}
		Node newHead=reverseRecursive(head.getNext());
		Node headNext=head.getNext();
		headNext.setNext(head);
		head.setNext(null);
		return newHead;
	}
	public void reverseIterative()
	{
		head=reverseIterative(head);
	}
	private Node reverseIterative(Node head)
	{
		Node currNode=head;
		Node prevNode=null;
		while(currNode!=null)
		{
			Node tempNode=currNode.getNext();
			currNode.setNext(prevNode);
			prevNode=currNode;
			currNode=tempNode;
		}
		return prevNode;
	}
}

File: Node/linkedList/singlyLinked/singlyLinkedLinkedList.js
class Node {
    constructor() {
        this.dataVal = null;
        this.next = null;
    }
    getDataVal() {
        return this.dataVal;
    }
    setDataVal(dataVal) {
        this.dataVal = dataVal;
    }
    getNext() {
        return this.next;
    }
    setNext(next) {
        this.next = next;
    }
}

class SinglyLinkedLinkedList {
    constructor() {
        this.head = null;
    }
    addHead(dataVal) {
        let tempNode = new Node();
        tempNode.setDataVal(dataVal);
        if (this.head == null) {
            this.head = tempNode;
        }
        else {
            tempNode.setNext(this.head);
            this.head = tempNode;
        }
    }
    add(dataVal) {
        let tempNode = new Node();
        tempNode.setDataVal(dataVal);
        if (this.head == null) {
            this.head = tempNode;
        }
        else {
            let tempHead = this.head;
            while (tempHead.getNext()) {
                tempHead = tempHead.getNext();
            }
            tempHead.setNext(tempNode);
        }
    }
    addAtIndex(dataVal, index) {
        let tempNode = new Node();
        tempNode.setDataVal(dataVal);
        if (index == 0) {
            this.addHead(dataVal);
        }
        else {
            let currIndex = 0;
            let tempHead = this.head;
            while (tempHead.getNext() && index != 1) {
                tempHead = tempHead.getNext();
                currIndex++;
                if (currIndex == index - 1) {
                    break;
                }
            }
            tempNode.setNext(tempHead.getNext());
            tempHead.setNext(tempNode);
        }
    }
    updateHead(dataVal) {
        this.head.setDataVal(dataVal);
    }
    updateLast(dataVal) {
        let tempHead = this.head;
        while (tempHead.getNext()) {
            tempHead = tempHead.getNext();
        }
        tempHead.setDataVal(dataVal);
    }
    updateAtIndex(dataVal, index) {
        let tempHead = this.head;
        let currIndex = 0;
        while (tempHead.getNext() && currIndex != index) {
            tempHead = tempHead.getNext();
            currIndex++;
        }
        if (index > currIndex) {
            console.log('Index out of bounds');
        }
        else {
            tempHead.setDataVal(dataVal);
        }
    }
    deleteHead() {
        this.head = this.head.getNext();
    }
    deleteLast() {
        let tempHead = this.head;
        while (tempHead.getNext().getNext()) {
            tempHead = tempHead.getNext();
        }
        tempHead.setNext(null);
    }
    deleteAtIndex(index) {
        if (index == 0) {
            this.deleteHead();
        }
        else {
            let tempHead = this.head;
            let currIndex = 0;
            while (tempHead.getNext().getNext() && index != 1) {
                tempHead = tempHead.getNext();
                currIndex++;
                if (currIndex == index - 1) {
                    break;
                }
            }
            if (index > currIndex && index != 1) {
                console.log("Out of Bounds");
            }
            else {
                tempHead.setNext(tempHead.getNext().getNext());
            }
        }
    }
    size() {
        let size = 0;
        if (this.head != null) {
            let tempHead = this.head;
            while (tempHead.getNext()) {
                tempHead = tempHead.getNext();
                size++;
            }
            size++;
        }
        return size;
    }
    reverseIterative() {
        let currNode = this.head;
        let prevNode = null;
        while (currNode != null) {
            let tempNode = currNode.getNext();
            currNode.setNext(prevNode);
            prevNode = currNode;
            currNode = tempNode;
        }
        this.head = prevNode;
    }
    reverseRecersive() {
        this.head = this.recursiveReverseCall(this.head);
    }
    recursiveReverseCall(head) {
        if (head == null || head.getNext() == null) {
            return head;
        }
        let newHead = this.recursiveReverseCall(head.getNext());
        let nextHead = head.getNext();
        nextHead.setNext(head);
        head.setNext(null);
        return newHead;
    }
    printAllValues() {
        let values = "";
        let tempHead = this.head;
        while (tempHead.getNext()) {
            values += tempHead.dataVal + " , ";
            tempHead = tempHead.getNext();
        }
        values += tempHead.dataVal;
        console.log(values);
    }
}

let sLL = new SinglyLinkedLinkedList();
let node = new Node();
sLL.addHead(10);
sLL.add(20);
sLL.add(30);
sLL.add(40);
sLL.addAtIndex(100, 4)
sLL.printAllValues();
sLL.updateHead(1.1)
sLL.updateAtIndex(3.3, 2)
sLL.updateLast(5.5);
sLL.printAllValues();
sLL.deleteHead();
sLL.deleteLast();
sLL.deleteAtIndex(1)
sLL.printAllValues();
console.log(sLL.size());
sLL.reverseIterative();
sLL.printAllValues();
sLL.reverseRecersive();
sLL.printAllValues();